### PR TITLE
add mirror service

### PR DESCRIPTION
**Pull request by mikehelmick**
Towards #928

## Proposed Changes

* Add mirror service that can mirror an index.txt / export archive structure from a CDN to the configured cloud storage. For use in key server migrations.

**Release Note**

```release-note
New: Mirror service `cmd/mirror` for mirroring an export directory from one key server to another.
```

### CODE REVIEW

File: internal/mirror/database/mirror.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+func (db *MirrorDB) Mirrors(ctx context.Context) ([]*model.Mirror, error) {
+	var mirrors []*model.Mirror
+
+	if err := db.db.InTx(ctx, pgx.ReadCommitted, func(tx pgx.Tx) error {
+		rows, err := tx.Query(ctx, `
+			SELECT
+				id, index_file, export_root, cloud_storage_bucket, filename_root
+			FROM
+				mirror
```
Comment (sethvargo) [inline internal/mirror/database/mirror.go:88]: `ORDER BY id` or something?

File: internal/mirror/database/mirror.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+					mirror_id = $1 AND filename = $2
+				`); err != nil {
+				return fmt.Errorf("failed to prepare delete DB statement: %w", err)
+			}
+
+			for _, mf := range toDelete {
+				if _, err := tx.Exec(ctx, deleteName, mf.MirrorID, mf.Filename); err != nil {
+					return fmt.Errorf("failed to delete mirrofile: %w", err)
+				}
+			}
```
Comment (sethvargo) [inline internal/mirror/database/mirror.go:153]: Do we need to check if the number of rows modified == len(toDelete)?

File: internal/mirror/database/mirror.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		}
+
+		// Create files if they still need to be created.
+		if len(wantFiles) > 0 {
+			if _, err := tx.Prepare(ctx, insertName, `
```
Comment (sethvargo) [inline internal/mirror/database/mirror.go:158]: This is the same transaction as the delete, right? If wantFiles and toDelete include the same (mirror_id, filename), what happens?

File: internal/mirror/mirror.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+func (s *Server) downloadIndex(mirror *model.Mirror) ([]string, error) {
+	client := &http.Client{
+		Timeout: s.config.IndexFileDownloadTimeout,
+	}
+	resp, err := client.Get(mirror.IndexFile)
+	if err != nil {
+		return nil, fmt.Errorf("error downloading index file: %w", err)
+	}
+	defer resp.Body.Close()
+	bytes, err := ioutil.ReadAll(resp.Body)
```
Comment (sethvargo) [inline internal/mirror/mirror.go:62]: ReadAll has no limit and will consume all available memory given. We should use a limitReader or something with a sane max here.

File: internal/mirror/mirror.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+			if err != nil {
+				if errors.Is(err, database.ErrAlreadyLocked) {
+					logger.Infow("mirror already locked, skipping", "mirrorID", mirror.ID)
+					continue
+				}
+				logger.Warn(err)
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			defer func() {
```
Comment (sethvargo) [inline internal/mirror/mirror.go:122]: This defer won't run until _all_ things in the loop have finished and the rest of the function finishes executing. I don't think that's what you want?

File: internal/mirror/mirror.go [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+			if err != nil {
+				logger.Errorw("unable to list mirror files", "error", err)
+				continue
+			}
+
+			// download the index
+			// curFiles contains the fullly qualified download link for the files.
+			curFiles, err := s.downloadIndex(mirror)
+			if err != nil {
+				logger.Errorw("unable to download index.txt", "error", err)
```
Comment (sethvargo) [inline internal/mirror/mirror.go:141]: continue or break?

File: internal/mirror/mirror.go [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+					}
+				}
+			}
+
+			// delete any files we no longer have
+			for filename, status := range actions {
+				if status.needsDelete() {
+					logger.Infow("deleting stale export file", "file", filename)
+
+					bsCtx, cancel := context.WithTimeout(ctx, time.Second*30)
```
Comment (sethvargo) [inline internal/mirror/mirror.go:172]: Nit: 30*time.Second

File: internal/mirror/mirror.go [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+			indexObjects := make([]*FileStatus, 0, len(actions))
+			// quickly bypass anything that doesn't need to be downloaded.
+			// This ensures that if we timeout, files we previously knew about remain saved.
+			for _, status := range actions {
+				if !status.needsDownload() {
+					status.Saved = true
+					indexObjects = append(indexObjects, status)
+				}
+			}
+			logger.Infow("mirror state", "mirrorID", mirror.ID, "total", len(actions), "existing", len(indexObjects), "todo", len(actions)-len(indexObjects))
```
Comment (sethvargo) [inline internal/mirror/mirror.go:194]: TODO

File: internal/mirror/mirror.go [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+					defer resp.Body.Close()
+					bytes, err := ioutil.ReadAll(resp.Body)
+					if err != nil {
+						logger.Errorw("unable to read file", "filename", filename, "error", err)
+						status.Failed = true
+						continue
+					}
+
+					// Write this file to the blob store.
+					bsCtx, cancel := context.WithTimeout(ctx, time.Minute)
```
Comment (sethvargo) [inline internal/mirror/mirror.go:225]: Why is the other 30*time.Second and this is one minute?

File: internal/mirror/mirror.go [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+		status := http.StatusOK
+		w.WriteHeader(status)
+		w.Write([]byte(http.StatusText(status)))
+	}
+}
+
+func deadlinePassed(deadline time.Time) bool {
+	return time.Now().After(deadline)
+}
```
Comment (sethvargo) [inline internal/mirror/mirror.go:276]: I'd really like to see this broken down into smaller functions that are easily tested. The current HTTP handler is very long and _very_ nested.

**Review by sethvargo — CHANGES_REQUESTED** [review]
I think the mirror _algorithm_ is fine, but the implementation has a few flaws. Since everything is a giant loop, none of the defers are running until the very end. Furthermore, for a very large N, we'll hit unnecessary timeouts due to lack of parallelization.

I'd like to suggest we:

1. Break apart the main HTTP handler into (at least) one testable function.

1. Leverage some parallelization (either primitively with goroutines or using a library like https://github.com/gammazero/workerpool) with a configurable max.

1. Add tests for `internal/mirror/mirror.go`. The blobstore should be stubbable to the in-memory implementation.

File: internal/mirror/database/mirror.go (hunk 1 above)
Comment (mikehelmick) [reply internal/mirror/database/mirror.go:88]: added, but not necessary for proper function.

File: internal/mirror/database/mirror.go (hunk 2 above)
Comment (mikehelmick) [reply internal/mirror/database/mirror.go:160]: added - it's individual row deletes on the prepared statement though

File: internal/mirror/database/mirror.go (hunk 3 above)
Comment (mikehelmick) [reply internal/mirror/database/mirror.go:166]: wantFiles only contains files that weren't read back into knownFiles. 

added comments to try and make this clearer

File: internal/mirror/mirror.go (hunk 4 above)
Comment (mikehelmick) [reply internal/mirror/mirror.go:63]: added limits to both downloads

File: internal/mirror/mirror.go (hunk 5 above)
Comment (mikehelmick) [reply internal/mirror/mirror.go]: extracted processing to a function so all defers will run.

File: internal/mirror/mirror.go (hunk 6 above)
Comment (mikehelmick) [reply internal/mirror/mirror.go]: continue.

File: internal/mirror/mirror.go (hunk 8 above)
Comment (mikehelmick) [reply internal/mirror/mirror.go]: toto - as in, what's left todo 🤣

File: internal/mirror/mirror.go (hunk 9 above)
Comment (mikehelmick) [reply internal/mirror/mirror.go]: one is delete and one is write

File: internal/mirror/mirror.go (hunk 10 above)
Comment (mikehelmick) [reply internal/mirror/mirror.go:290]: can do as a follow up if that's OK

**Comment by mikehelmick** [comment]
> I think the mirror _algorithm_ is fine, but the implementation has a few flaws. Since everything is a giant loop, none of the defers are running until the very end. Furthermore, for a very large N, we'll hit unnecessary timeouts due to lack of parallelization.

Parallelization is achieved at the HTTP layer - will have a cron to active this service every 60 seconds. if one mirror is locked and taking a while, the next invocation will move onto the next one.

State is saved on an http request timeout and processing will continue on next invoke.


> 
> I'd like to suggest we:
> 
> 1. Break apart the main HTTP handler into (at least) one testable function.

will do in follow up.

> 2. Leverage some parallelization (either primitively with goroutines or using a library like https://github.com/gammazero/workerpool) with a configurable max.

we can add additional parallelism, but doesn't seem to be necessary in the testing i've done.

> 3. Add tests for `internal/mirror/mirror.go`. The blobstore should be stubbable to the in-memory implementation.

would like to do in follow up

File: internal/mirror/mirror.go (hunk 10 above)
Comment (sethvargo) [reply internal/mirror/mirror.go:289]: yup

**Review by sethvargo — APPROVED** [review]
/lgtm
/hold

in case someone else wants to give it a once-over. Feel free to unhold.

**Pull request merged**